src/main/java/com/uber/cadence/internal/metrics/ReplayAwareScope.java

```diff
@@ -0,0 +1,183 @@
+/*
+ *  Copyright 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *  Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"). You may not
+ *  use this file except in compliance with the License. A copy of the License is
+ *  located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file. This file is distributed on
+ *  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+package com.uber.cadence.internal.metrics;
+
+import com.uber.cadence.internal.replay.ReplayAware;
+import com.uber.m3.tally.Buckets;
+import com.uber.m3.tally.Capabilities;
+import com.uber.m3.tally.Counter;
+import com.uber.m3.tally.Gauge;
+import com.uber.m3.tally.Histogram;
+import com.uber.m3.tally.Scope;
+import com.uber.m3.tally.ScopeCloseException;
+import com.uber.m3.tally.Stopwatch;
+import com.uber.m3.tally.Timer;
+import com.uber.m3.util.Duration;
+import java.util.Map;
+import java.util.Objects;
+
+public class ReplayAwareScope implements Scope {
+  Scope scope;
+  ReplayAware context;
+
+  public ReplayAwareScope(Scope scope, ReplayAware context) {
+    this.scope = Objects.requireNonNull(scope);
+    this.context = Objects.requireNonNull(context);
+  }
+
+  private static class ReplayAwareCounter implements Counter {
+    Counter counter;
+    ReplayAware context;
+
+    public ReplayAwareCounter(Counter counter, ReplayAware context) {
+      this.counter = Objects.requireNonNull(counter);
+      this.context = Objects.requireNonNull(context);
+    }
+
+    @Override
+    public void inc(long delta) {
+      if (context.isReplaying()) {
+        return;
+      }
+
+      counter.inc(delta);
+    }
+  }
+
+  private static class ReplayAwareGauge implements Gauge {
+    Gauge gauge;
+    ReplayAware context;
+
+    public ReplayAwareGauge(Gauge gauge, ReplayAware context) {
+      this.gauge = Objects.requireNonNull(gauge);
+      this.context = Objects.requireNonNull(context);
+    }
+
+    @Override
+    public void update(double value) {
+      if (context.isReplaying()) {
+        return;
+      }
+
+      gauge.update(value);
+    }
+  }
+
+  private static class ReplayAwareTimer implements Timer {
+    Timer timer;
+    ReplayAware context;
+
+    public ReplayAwareTimer(Timer timer, ReplayAware context) {
+      this.timer = Objects.requireNonNull(timer);
```

Comment:
This is not going to work. The problem is that this implementation reuses the timer from the wrapped scope. And the TimerImpl.start calls into the system clock. So during replay the start time will be the time of the replay, not the time the start was called during the first execution. 
Make sure that Workflow.currentTimeMillis() is used everywhere when calculating latencies.

Reply:
Fixed.